controllers/namespacescope/namespacescope_controller.go

```diff
@@ -0,0 +1,250 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package namespacescope
+
+import (
+	"context"
+	"time"
+
+	gset "github.com/deckarep/golang-set"
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/discovery"
+	"k8s.io/klog"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	nssv1 "github.com/IBM/ibm-namespace-scope-operator/api/v1"
+
+	operatorv1alpha1 "github.com/IBM/operand-deployment-lifecycle-manager/api/v1alpha1"
+	"github.com/IBM/operand-deployment-lifecycle-manager/controllers/constant"
+	deploy "github.com/IBM/operand-deployment-lifecycle-manager/controllers/operator"
+	"github.com/IBM/operand-deployment-lifecycle-manager/controllers/util"
+)
+
+// Reconciler automagically updates NamespaceScope CR with the proporly namespace
+type Reconciler struct {
+	*deploy.ODLMOperator
+}
+
+// ReconcileOperandRequest reads that state of the cluster for OperandRequest object and update NamespaceScope CR based on the state read
+func (r *Reconciler) ReconcileOperandRequest(req ctrl.Request) (_ ctrl.Result, reconcileErr error) {
+	// Creat context for the OperandBindInfo reconciler
+	ctx := context.Background()
+
+	exist, err := r.checkNamespaceScapeAPI()
+	if err != nil {
+		return ctrl.Result{}, err
+	} else if !exist {
+		return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+	}
+
+	nss, err := r.getNamespaceScopeCR(ctx)
+	if err != nil {
+		klog.Error(err)
+		return ctrl.Result{}, err
+	} else if nss == nil {
+		klog.Warning("Not found NamespaceScope instance, ignore update it.")
+		return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+	}
+
+	// Fetch the OperandRequest instance
+	requestInstance := &operatorv1alpha1.OperandRequest{}
+	if err := r.Client.Get(ctx, req.NamespacedName, requestInstance); err != nil {
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	originalNss := nss.DeepCopy()
+
+	var nsMems []string
+
+	defer func() {
+		if !util.StringSliceContentEqual(nsMems, nss.Spec.NamespaceMembers) {
+			nss.Spec.NamespaceMembers = nsMems
+			if err := r.Patch(ctx, nss, client.MergeFrom(originalNss)); err != nil {
+				reconcileErr = errors.Wrapf(err, "failed to update NamespaceScope %s/%s", nss.Namespace, nss.Name)
+			}
+			klog.V(2).Infof("Updated NamespaceScope %s/%s", nss.Namespace, nss.Name)
+		}
+	}()
+
+	// Remove finalizer when DeletionTimestamp none zero
+	if !requestInstance.ObjectMeta.DeletionTimestamp.IsZero() {
+		// Check and remove namespaceMember from NamespaceScope CR
+		nsMems, err = r.removeNamespaceMemberFromNamespaceScope(ctx, req)
```

Comment:
I think the function name `removeNamespaceMemberFromNamespaceScope` did not make sense here.

From my understand, It fetches the current namespace from the list of `OperandRequest`, which has the same functionality with `r. getOpreqNs`, except that it doesn't include the namespace of current `OperanRequest`.

Mu though would be that we create a new function call `r.UpdateNamespaceMemberInNamespaceScope`. It includes `r.getOpreqNs` and `r. r.getOpregNs`, then returns the `nsMems` like what we did in line 99-125. and we call it at line 87, before `DeletionTimestamp is none zero`.

And if we need to `removeNamespaceMemberFromNamespaceScope`, we could simply remove the current namespace of OperandRequest from the `nsMems` set. then the defer function could Update the NSS CR. Also the code from line 99-125 would be unnecessary.

Reply:
It's a good idea

Reply:
Another point I missed is we can't delete the namespace from the namespacescope CR before the operandrequest is deleted. Otherwise ODLM won't has permission in the namespace of operandrequest to clean up resources

Reply:
Ah! So NSS controller needs to hold the updates in NSS CR until the `OperandRequest` in that ns is deleted. Great find!